pkg/estimator/client/service.go

```diff
@@ -0,0 +1,69 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+)
+
+// ResolveCluster parses Service resource content by itself.
+// Fixes Issue https://github.com/karmada-io/karmada/issues/2487
+// Modified from "k8s.io/apiserver/pkg/util/proxy/proxy.go:92 => func ResolveCluster"
+func resolveCluster(kubeClient kubernetes.Interface, namespace, id string, port int32) (string, error) {
+	svc, err := kubeClient.CoreV1().Services(namespace).Get(context.TODO(), id, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			/*
+			 * When Deploying Karmada in Host Kubernetes Cluster, the kubeClient will connect kube-apiserver
+			 * of Karmada Control Plane, rather than of host cluster.
+			 * But the Service resource is defined in Host Kubernetes Cluster. So we cannot get its content here.
+			 * The best thing we can do is just glue host:port together, and try to connect to it.
+			 */
+			return net.JoinHostPort(id, fmt.Sprintf("%d", port)), nil
```

Comment:
Seems like now we resolve the DNS by ourselves instead of CoreDNS, right? For some scenarios without deploying CoreDNS, this patch may be helpful. I still have some questions.

The parameter `kubeClient` is corresponding with karmada-apiserver instead of host cluster. So normally we can't find any service here. Did you create a external service in karmada control plane?

BTW, if so, I think only external service will take effect. A `clusterIP` or `nodePort` service created in karmada control plane will never be connected cause the IP is fake.

Reply:
(1) 

> The parameter kubeClient is corresponding with karmada-apiserver instead of host cluster. So normally we can't find any service here. Did you create a external service in karmada control plane?

Yes, when I need to deploy karmada-scheduler without a host cluster,  I create a Service in karamda-apiserver.

(2) 

> BTW, if so, I think only external service will take effect. A clusterIP or nodePort service created in karmada control plane will never be connected cause the IP is fake.

Yes, I use create a Service with ExternalName type.

(3) 

The commented logic is for scheduler-estimator deployed in host cluster. Because karamda-apiserver don't have the required `Service`, I cannot parse `Service` content by myself. I fallback to the original logic, aka just try to connect to `host:port`.